File: source/SharpFlame/INI.cs
using System;
using System.IO;
using Microsoft.VisualBasic;
using SharpFlame.Collections;

namespace SharpFlame
{
    public class clsINIRead
    {
        public class clsSection
        {
            public string Name;
#if !Mono
            public struct sProperty
            {
#else
				public class sProperty
				{
#endif
                public string Name;
                public string Value;
#if !Mono
            }
#else
			}
#endif
            public SimpleList<sProperty> Properties = new SimpleList<sProperty>();

            public void CreateProperty(string Name, string Value)
            {
                sProperty NewProperty = new sProperty();

                NewProperty.Name = Name;
                NewProperty.Value = Value;
                Properties.Add(NewProperty);
            }

            public clsResult ReadFile(StreamReader File)
            {
                clsResult ReturnResult = new clsResult("");

                int InvalidLineCount = 0;
                int CurrentEntryNum = -1;
                string LineText = null;
                int A = 0;

                do
                {
                    LineText = File.ReadLine();
                    if ( LineText == null )
                    {
                        break;
                    }
                    LineText = LineText.Trim();
                    A = LineText.IndexOf('#');
                    if ( A >= 0 )
                    {
                        LineText = Strings.Left(LineText, A).Trim();
                    }
                    if ( LineText.Length >= 1 )
                    {
                        A = LineText.IndexOf('=');
                        if ( A >= 0 )
                        {
                            CreateProperty(LineText.Substring(0, A).ToLower().Trim(), LineText.Substring(A + 1, LineText.Length - A - 1).Trim());
                        }
                        else
                        {
                            InvalidLineCount++;
                        }
                    }
                    else if ( LineText.Length > 0 )
                    {
                        InvalidLineCount++;
                    }
                } while ( true );

                Properties.RemoveBuffer();

                if ( InvalidLineCount > 0 )
                {
                    ReturnResult.WarningAdd("There were " + Convert.ToString(InvalidLineCount) + " invalid lines that were ignored.");
                }

                return ReturnResult;
            }

            public clsResult Translate(int SectionNum, clsSectionTranslator Translator, sErrorCount ErrorCount)
            {
                clsResult ReturnResult = new clsResult("Section " + Name);

                int A = 0;
                enumTranslatorResult TranslatorResult = default(enumTranslatorResult);

                for ( A = 0; A <= Properties.Count - 1; A++ )
                {
                    TranslatorResult = Translator.Translate(SectionNum, Properties[A]);
                    switch ( TranslatorResult )
                    {
                        case enumTranslatorResult.NameUnknown:
                            if ( ErrorCount.NameErrorCount < 16 )
                            {
                                ReturnResult.WarningAdd("Property name " + Convert.ToString(ControlChars.Quote) + Properties[A].Name +
                                                        Convert.ToString(ControlChars.Quote) + " is unknown.");
                            }
                            ErrorCount.NameErrorCount++;
                            break;
                        case enumTranslatorResult.ValueInvalid:
                            if ( ErrorCount.ValueErrorCount < 16 )
                            {
                                ReturnResult.WarningAdd("Value " + Convert.ToString(ControlChars.Quote) + Properties[A].Value +
                                                        Convert.ToString(ControlChars.Quote) + " for property name " +
                                                        Convert.ToString(ControlChars.Quote) + Properties[A].Name +
                                                        Convert.ToString(ControlChars.Quote) + " is not valid.");
                            }
                            ErrorCount.ValueErrorCount++;
                            break;
                    }
                }

                return ReturnResult;
            }

            public clsResult Translate(clsTranslator Translator)
            {
                clsResult ReturnResult = new clsResult("Section " + Name);

                int A = 0;
                enumTranslatorResult TranslatorResult = default(enumTranslatorResult);
                sErrorCount ErrorCount = new sErrorCount();

                ErrorCount.NameWarningCountMax = 16;
                ErrorCount.ValueWarningCountMax = 16;

                for ( A = 0; A <= Properties.Count - 1; A++ )
                {
                    TranslatorResult = Translator.Translate(Properties[A]);
                    switch ( TranslatorResult )
                    {
                        case enumTranslatorResult.NameUnknown:
                            if ( ErrorCount.NameErrorCount < 16 )
                            {
                                ReturnResult.WarningAdd("Property name " + Convert.ToString(ControlChars.Quote) + Properties[A].Name +
                                                        Convert.ToString(ControlChars.Quote) + " is unknown.");
                            }
                            ErrorCount.NameErrorCount++;
                            break;
                        case enumTranslatorResult.ValueInvalid:
                            if ( ErrorCount.ValueErrorCount < 16 )
                            {
                                ReturnResult.WarningAdd("Value " + Convert.ToString(ControlChars.Quote) + Properties[A].Value +
                                                        Convert.ToString(ControlChars.Quote) + " for property name " +
                                                        Convert.ToString(ControlChars.Quote) + Properties[A].Name +
                                                        Convert.ToString(ControlChars.Quote) + " is not valid.");
                            }
                            ErrorCount.ValueErrorCount++;
                            break;
                    }
                }

                if ( ErrorCount.NameErrorCount > ErrorCount.NameWarningCountMax )
                {
                    ReturnResult.WarningAdd("There were " + Convert.ToString(ErrorCount.NameErrorCount) + " unknown property names that were ignored.");
                }
                if ( ErrorCount.ValueErrorCount > ErrorCount.ValueWarningCountMax )
                {
                    ReturnResult.WarningAdd("There were " + Convert.ToString(ErrorCount.ValueErrorCount) + " invalid values that were ignored.");
                }

                return ReturnResult;
            }

            public string GetLastPropertyValue(string LCasePropertyName)
            {
                int A = 0;

                for ( A = Properties.Count - 1; A >= 0; A-- )
                {
                    if ( Properties[A].Name == LCasePropertyName )
                    {
                        return Properties[A].Value;
                    }
                }
                return null;
            }
        }

        public SimpleList<clsSection> Sections = new SimpleList<clsSection>();
        public clsSection RootSection;

        public void CreateSection(string Name)
        {
            clsSection newSection = new clsSection();
            newSection.Name = Name;

            Sections.Add(newSection);
        }

        public clsResult ReadFile(StreamReader File)
        {
            clsResult ReturnResult = new clsResult("Reading INI");

            int InvalidLineCount = 0;
            int CurrentEntryNum = -1;
            string LineText = null;
            int A = 0;
            string SectionName = "";

            RootSection = new clsSection();

            do
            {
                LineText = File.ReadLine();
                if ( LineText == null )
                {
                    break;
                }
                LineText = LineText.Trim();
                A = LineText.IndexOf('#');
                if ( A >= 0 )
                {
                    LineText = Strings.Left(LineText, A).Trim();
                }
                if ( LineText.Length >= 2 )
                {
                    if ( LineText[0] == '[' )
                    {
                        if ( LineText[LineText.Length - 1] == ']' )
                        {
                            SectionName = LineText.Substring(1, LineText.Length - 2);
                            for ( A = 0; A <= Sections.Count - 1; A++ )
                            {
                                if ( Sections[A].Name == SectionName )
                                {
                                    break;
                                }
                            }
                            CurrentEntryNum = A;
                            if ( CurrentEntryNum == Sections.Count )
                            {
                                CreateSection(SectionName);
                            }
                        }
                        else
                        {
                            InvalidLineCount++;
                        }
                    }
                    else if ( CurrentEntryNum >= 0 )
                    {
                        A = LineText.IndexOf('=');
                        if ( A >= 0 )
                        {
                            Sections[CurrentEntryNum].CreateProperty(LineText.Substring(0, A).ToLower().Trim(),
                                LineText.Substring(A + 1, LineText.Length - A - 1).Trim());
                        }
                        else
                        {
                            InvalidLineCount++;
                        }
                    }
                    else
                    {
                        A = LineText.IndexOf('=');
                        if ( A >= 0 )
                        {
                            RootSection.CreateProperty(LineText.Substring(0, A).ToLower().Trim(), LineText.Substring(A + 1, LineText.Length - A - 1).Trim());
                        }
                        else
                        {
                            InvalidLineCount++;
                        }
                    }
                }
                else if ( LineText.Length > 0 )
                {
                    InvalidLineCount++;
                }
            } while ( true );

            Sections.RemoveBuffer();

            if ( InvalidLineCount > 0 )
            {
                ReturnResult.WarningAdd("There were " + Convert.ToString(InvalidLineCount) + " invalid lines that were ignored.");
            }

            return ReturnResult;
        }

        public enum enumTranslatorResult
        {
            NameUnknown,
            ValueInvalid,
            Translated
        }

        public struct sErrorCount
        {
            public int NameErrorCount;
            public int ValueErrorCount;
            public int NameWarningCountMax;
            public int ValueWarningCountMax;
        }

        public clsResult Translate(clsSectionTranslator Translator)
        {
            clsResult ReturnResult = new clsResult("Translating INI");

            int A = 0;
            sErrorCount ErrorCount = new sErrorCount();

            ErrorCount.NameWarningCountMax = 16;
            ErrorCount.ValueWarningCountMax = 16;

            for ( A = 0; A <= Sections.Count - 1; A++ )
            {
                ReturnResult.Add(Sections[A].Translate(A, Translator, ErrorCount));
            }

            if ( ErrorCount.NameErrorCount > ErrorCount.NameWarningCountMax )
            {
                ReturnResult.WarningAdd("There were " + Convert.ToString(ErrorCount.NameErrorCount) + " unknown property names that were ignored.");
            }
            if ( ErrorCount.ValueErrorCount > ErrorCount.ValueWarningCountMax )
            {
                ReturnResult.WarningAdd("There were " + Convert.ToString(ErrorCount.ValueErrorCount) + " invalid values that were ignored.");
            }

            return ReturnResult;
        }

        public abstract class clsSectionTranslator
        {
            public abstract enumTranslatorResult Translate(int INISectionNum, clsSection.sProperty INIProperty);
        }

        public abstract class clsTranslator
        {
            public abstract enumTranslatorResult Translate(clsSection.sProperty INIProperty);
        }
    }

    public class clsINIWrite
    {
        public StreamWriter File;
        public char LineEndChar;
        public char EqualsChar = '=';

        public clsINIWrite()
        {
            LineEndChar = '\n';
        }

        public static clsINIWrite CreateFile(Stream Output)
        {
            clsINIWrite NewINI = new clsINIWrite();

            NewINI.File = new StreamWriter(Output, App.UTF8Encoding);

            return NewINI;
        }

        public void SectionName_Append(string Name)
        {
            Name = Name.Replace(LineEndChar.ToString(), "");

            File.Write('[' + Name + Convert.ToString(']') + Convert.ToString(LineEndChar));
        }

        public void Property_Append(string Name, string Value)
        {
            Name = Name.Replace(LineEndChar.ToString(), "");
            Name = Name.Replace(EqualsChar.ToString(), "");
            Value = Value.Replace(LineEndChar.ToString(), "");

            File.Write(Name + " " + Convert.ToString(EqualsChar) + " " + Value + Convert.ToString(LineEndChar));
        }

        public void Gap_Append()
        {
            File.Write(LineEndChar);
        }
    }
}

File: reference/FlaMEVB/INI.vb

Public Class clsINIRead
    Public Class clsSection
        Public Name As String
#If Not Mono Then
        Public Structure sProperty
#Else
        Public Class sProperty
#End If
            Public Name As String
            Public Value As String
#If Not Mono Then
        End Structure
#Else
        End Class
#End If
        Public Properties As New SimpleList(Of sProperty)

        Public Sub CreateProperty(Name As String, Value As String)
            Dim NewProperty As New sProperty

            NewProperty.Name = Name
            NewProperty.Value = Value
            Properties.Add(NewProperty)
        End Sub

        Public Function ReadFile(File As System.IO.StreamReader) As clsResult
            Dim ReturnResult As New clsResult("")

            Dim InvalidLineCount As Integer = 0
            Dim CurrentEntryNum As Integer = -1
            Dim LineText As String = Nothing
            Dim A As Integer

            Do
                LineText = File.ReadLine
                If LineText Is Nothing Then
                    Exit Do
                End If
                LineText = LineText.Trim
                A = LineText.IndexOf("#"c)
                If A >= 0 Then
                    LineText = Strings.Left(LineText, A).Trim
                End If
                If LineText.Length >= 1 Then
                    A = LineText.IndexOf("="c)
                    If A >= 0 Then
                        CreateProperty(LineText.Substring(0, A).ToLower.Trim, LineText.Substring(A + 1, LineText.Length - A - 1).Trim)
                    Else
                        InvalidLineCount += 1
                    End If
                ElseIf LineText.Length > 0 Then
                    InvalidLineCount += 1
                End If
            Loop

            Properties.RemoveBuffer()

            If InvalidLineCount > 0 Then
                ReturnResult.WarningAdd("There were " & InvalidLineCount & " invalid lines that were ignored.")
            End If

            Return ReturnResult
        End Function

        Public Function Translate(SectionNum As Integer, Translator As clsINIRead.clsSectionTranslator, ByRef ErrorCount As sErrorCount) As clsResult
            Dim ReturnResult As New clsResult("Section " & Name)

            Dim A As Integer
            Dim TranslatorResult As enumTranslatorResult

            For A = 0 To Properties.Count - 1
                TranslatorResult = Translator.Translate(SectionNum, Properties(A))
                Select Case TranslatorResult
                    Case enumTranslatorResult.NameUnknown
                        If ErrorCount.NameErrorCount < 16 Then
                            ReturnResult.WarningAdd("Property name " & ControlChars.Quote & Properties(A).Name & ControlChars.Quote & " is unknown.")
                        End If
                        ErrorCount.NameErrorCount += 1
                    Case enumTranslatorResult.ValueInvalid
                        If ErrorCount.ValueErrorCount < 16 Then
                            ReturnResult.WarningAdd("Value " & ControlChars.Quote & Properties(A).Value & ControlChars.Quote & " for property name " & ControlChars.Quote & Properties(A).Name & ControlChars.Quote & " is not valid.")
                        End If
                        ErrorCount.ValueErrorCount += 1
                End Select
            Next

            Return ReturnResult
        End Function

        Public Function Translate(Translator As clsINIRead.clsTranslator) As clsResult
            Dim ReturnResult As New clsResult("Section " & Name)

            Dim A As Integer
            Dim TranslatorResult As enumTranslatorResult
            Dim ErrorCount As sErrorCount

            ErrorCount.NameWarningCountMax = 16
            ErrorCount.ValueWarningCountMax = 16

            For A = 0 To Properties.Count - 1
                TranslatorResult = Translator.Translate(Properties(A))
                Select Case TranslatorResult
                    Case enumTranslatorResult.NameUnknown
                        If ErrorCount.NameErrorCount < 16 Then
                            ReturnResult.WarningAdd("Property name " & ControlChars.Quote & Properties(A).Name & ControlChars.Quote & " is unknown.")
                        End If
                        ErrorCount.NameErrorCount += 1
                    Case enumTranslatorResult.ValueInvalid
                        If ErrorCount.ValueErrorCount < 16 Then
                            ReturnResult.WarningAdd("Value " & ControlChars.Quote & Properties(A).Value & ControlChars.Quote & " for property name " & ControlChars.Quote & Properties(A).Name & ControlChars.Quote & " is not valid.")
                        End If
                        ErrorCount.ValueErrorCount += 1
                End Select
            Next

            If ErrorCount.NameErrorCount > ErrorCount.NameWarningCountMax Then
                ReturnResult.WarningAdd("There were " & ErrorCount.NameErrorCount & " unknown property names that were ignored.")
            End If
            If ErrorCount.ValueErrorCount > ErrorCount.ValueWarningCountMax Then
                ReturnResult.WarningAdd("There were " & ErrorCount.ValueErrorCount & " invalid values that were ignored.")
            End If

            Return ReturnResult
        End Function

        Public Function GetLastPropertyValue(LCasePropertyName As String) As String
            Dim A As Integer

            For A = Properties.Count - 1 To 0 Step -1
                If Properties(A).Name = LCasePropertyName Then
                    Return Properties(A).Value
                End If
            Next
            Return Nothing
        End Function
    End Class
    Public Sections As New SimpleList(Of clsSection)
    Public RootSection As clsSection

    Public Sub CreateSection(Name As String)
        Dim newSection As New clsSection
        newSection.Name = Name

        Sections.Add(newSection)
    End Sub

    Public Function ReadFile(File As IO.StreamReader) As clsResult
        Dim ReturnResult As New clsResult("Reading INI")

        Dim InvalidLineCount As Integer = 0
        Dim CurrentEntryNum As Integer = -1
        Dim LineText As String = Nothing
        Dim A As Integer
        Dim SectionName As String

        RootSection = New clsSection

        Do
            LineText = File.ReadLine
            If LineText Is Nothing Then
                Exit Do
            End If
            LineText = LineText.Trim
            A = LineText.IndexOf("#"c)
            If A >= 0 Then
                LineText = Strings.Left(LineText, A).Trim
            End If
            If LineText.Length >= 2 Then
                If LineText.Chars(0) = "["c Then
                    If LineText.Chars(LineText.Length - 1) = "]"c Then
                        SectionName = LineText.Substring(1, LineText.Length - 2)
                        For A = 0 To Sections.Count - 1
                            If Sections(A).Name = SectionName Then
                                Exit For
                            End If
                        Next
                        CurrentEntryNum = A
                        If CurrentEntryNum = Sections.Count Then
                            CreateSection(SectionName)
                        End If
                    Else
                        InvalidLineCount += 1
                    End If
                ElseIf CurrentEntryNum >= 0 Then
                    A = LineText.IndexOf("="c)
                    If A >= 0 Then
                        Sections(CurrentEntryNum).CreateProperty(LineText.Substring(0, A).ToLower.Trim, LineText.Substring(A + 1, LineText.Length - A - 1).Trim)
                    Else
                        InvalidLineCount += 1
                    End If
                Else
                    A = LineText.IndexOf("="c)
                    If A >= 0 Then
                        RootSection.CreateProperty(LineText.Substring(0, A).ToLower.Trim, LineText.Substring(A + 1, LineText.Length - A - 1).Trim)
                    Else
                        InvalidLineCount += 1
                    End If
                End If
            ElseIf LineText.Length > 0 Then
                InvalidLineCount += 1
            End If
        Loop

        Sections.RemoveBuffer()

        If InvalidLineCount > 0 Then
            ReturnResult.WarningAdd("There were " & InvalidLineCount & " invalid lines that were ignored.")
        End If

        Return ReturnResult
    End Function

    Public Enum enumTranslatorResult As Byte
        NameUnknown
        ValueInvalid
        Translated
    End Enum

    Public Structure sErrorCount
        Public NameErrorCount As Integer
        Public ValueErrorCount As Integer
        Public NameWarningCountMax As Integer
        Public ValueWarningCountMax As Integer
    End Structure

    Public Function Translate(Translator As clsINIRead.clsSectionTranslator) As clsResult
        Dim ReturnResult As New clsResult("Translating INI")

        Dim A As Integer
        Dim ErrorCount As sErrorCount

        ErrorCount.NameWarningCountMax = 16
        ErrorCount.ValueWarningCountMax = 16

        For A = 0 To Sections.Count - 1
            ReturnResult.Add(Sections(A).Translate(A, Translator, ErrorCount))
        Next

        If ErrorCount.NameErrorCount > ErrorCount.NameWarningCountMax Then
            ReturnResult.WarningAdd("There were " & ErrorCount.NameErrorCount & " unknown property names that were ignored.")
        End If
        If ErrorCount.ValueErrorCount > ErrorCount.ValueWarningCountMax Then
            ReturnResult.WarningAdd("There were " & ErrorCount.ValueErrorCount & " invalid values that were ignored.")
        End If

        Return ReturnResult
    End Function

    Public MustInherit Class clsSectionTranslator

        Public MustOverride Function Translate(INISectionNum As Integer, INIProperty As clsINIRead.clsSection.sProperty) As enumTranslatorResult
    End Class

    Public MustInherit Class clsTranslator

        Public MustOverride Function Translate(INIProperty As clsINIRead.clsSection.sProperty) As enumTranslatorResult
    End Class
End Class

Public Class clsINIWrite

    Public File As IO.StreamWriter
    Public LineEndChar As Char = Chr(10)
    Public EqualsChar As Char = "="c

    Public Sub New()

    End Sub

    Public Shared Function CreateFile(Output As IO.Stream) As clsINIWrite
        Dim NewINI As New clsINIWrite

        NewINI.File = New IO.StreamWriter(Output, UTF8Encoding)

        Return NewINI
    End Function

    Public Sub SectionName_Append(Name As String)

        Name = Name.Replace(LineEndChar, "")

        File.Write("["c & Name & "]"c & LineEndChar)
    End Sub

    Public Sub Property_Append(Name As String, Value As String)

        Name = Name.Replace(LineEndChar, "")
        Name = Name.Replace(EqualsChar, "")
        Value = Value.Replace(LineEndChar, "")

        File.Write(Name & " " & EqualsChar & " " & Value & LineEndChar)
    End Sub

    Public Sub Gap_Append()

        File.Write(LineEndChar)
    End Sub
End Class
